src/libcore/slice/mod.rs

```diff
@@ -1243,59 +1228,47 @@ macro_rules! iterator {
                 }
             }
 
-            fn rfind<F>(&mut self, mut predicate: F) -> Option<Self::Item>
-                where F: FnMut(&Self::Item) -> bool,
-            {
-                self.rsearch_while(None, move |elt| {
-                    if predicate(&elt) {
-                        SearchWhile::Done(Some(elt))
-                    } else {
-                        SearchWhile::Continue
-                    }
-                })
-            }
-
-        }
-
-        // search_while is a generalization of the internal iteration methods.
-        impl<'a, T> $name<'a, T> {
-            // search through the iterator's element using the closure `g`.
-            // if no element was found, return `default`.
-            fn search_while<Acc, G>(&mut self, default: Acc, mut g: G) -> Acc
-                where Self: Sized,
-                      G: FnMut($elem) -> SearchWhile<Acc>
+            #[inline]
+            fn try_rfold<B, F, R>(&mut self, init: B, mut f: F) -> R where
+                Self: Sized, F: FnMut(B, Self::Item) -> R, R: Try<Ok=B>
             {
                 // manual unrolling is needed when there are conditional exits from the loop
+                let mut accum = init;
                 unsafe {
+                    if mem::size_of::<T>() != 0 {
+                        assume(!self.ptr.is_null());
+                        assume(self.ptr <= self.end);
+                    }
```

Comment:
What's the reason these assumes were added here?

Reply:
`next` and `next_back` have assumptions, so I put them here too.  Given their status as arcana (from the `rposition` PR), I could certainly remove them...

Reply:
It's a shame if they should be needed, but I know they are needed for *some* optimizations in next. We should probably have some evidence for each of them being added. The non-null one is at least used when we create an `Option<&T>`, and that doesn't happen in this method, so it doesn't seem necessary?

Reply:
Hmm, `search_while` didn't have them, so hopefully they're not needed. I'll remove.